File: typescript/src/shared/tools.ts
import {z} from 'zod';

import {
  createCustomerPrompt,
  listCustomersPrompt,
} from '@/shared/customers/prompts';
import {
  createProductPrompt,
  listProductsPrompt,
} from '@/shared/products/prompts';
import {createPricePrompt, listPricesPrompt} from '@/shared/prices/prompts';
import {createPaymentLinkPrompt} from '@/shared/paymentLinks/prompts';
import {
  createInvoicePrompt,
  listInvoicesPrompt,
  finalizeInvoicePrompt,
} from '@/shared/invoices/prompts';
import {createInvoiceItemPrompt} from '@/shared/invoiceItems/prompts';
import {retrieveBalancePrompt} from '@/shared/balance/prompts';
import {createRefundPrompt} from '@/shared/refunds/prompts';
import {searchDocumentationPrompt} from '@/shared/documentation/prompts';
import {listPaymentIntentsPrompt} from '@/shared/paymentIntents/prompts';
import {listSubscriptionsPrompt} from '@/shared/subscriptions/prompts';

import {
  createCustomerParameters,
  listCustomersParameters,
} from '@/shared/customers/parameters';
import {
  createProductParameters,
  listProductsParameters,
} from '@/shared/products/parameters';
import {
  createPriceParameters,
  listPricesParameters,
} from '@/shared/prices/parameters';
import {createPaymentLinkParameters} from '@/shared/paymentLinks/parameters';
import {
  createInvoiceParameters,
  listInvoicesParameters,
  finalizeInvoiceParameters,
} from '@/shared/invoices/parameters';
import {createInvoiceItemParameters} from '@/shared/invoiceItems/parameters';
import {retrieveBalanceParameters} from '@/shared/balance/parameters';
import {createRefundParameters} from '@/shared/refunds/parameters';
import {searchDocumentationParameters} from '@/shared/documentation/parameters';
import {listPaymentIntentsParameters} from '@/shared/paymentIntents/parameters';
import {listSubscriptionsParameters} from '@/shared/subscriptions/parameters';

import type {Context} from './configuration';

export type Tool = {
  method: string;
  name: string;
  description: string;
  parameters: z.ZodObject<any, any, any, any>;
  actions: {
    [key: string]: {
      [action: string]: boolean;
    };
  };
};

const tools = (context: Context): Tool[] => [
  {
    method: 'create_customer',
    name: 'Create Customer',
    description: createCustomerPrompt(context),
    parameters: createCustomerParameters(context),
    actions: {
      customers: {
        create: true,
      },
    },
  },
  {
    method: 'list_customers',
    name: 'List Customers',
    description: listCustomersPrompt(context),
    parameters: listCustomersParameters(context),
    actions: {
      customers: {
        read: true,
      },
    },
  },
  {
    method: 'create_product',
    name: 'Create Product',
    description: createProductPrompt(context),
    parameters: createProductParameters(context),
    actions: {
      products: {
        create: true,
      },
    },
  },
  {
    method: 'list_products',
    name: 'List Products',
    description: listProductsPrompt(context),
    parameters: listProductsParameters(context),
    actions: {
      products: {
        read: true,
      },
    },
  },
  {
    method: 'create_price',
    name: 'Create Price',
    description: createPricePrompt(context),
    parameters: createPriceParameters(context),
    actions: {
      prices: {
        create: true,
      },
    },
  },
  {
    method: 'list_prices',
    name: 'List Prices',
    description: listPricesPrompt(context),
    parameters: listPricesParameters(context),
    actions: {
      prices: {
        read: true,
      },
    },
  },
  {
    method: 'create_payment_link',
    name: 'Create Payment Link',
    description: createPaymentLinkPrompt(context),
    parameters: createPaymentLinkParameters(context),
    actions: {
      paymentLinks: {
        create: true,
      },
    },
  },
  {
    method: 'create_invoice',
    name: 'Create Invoice',
    description: createInvoicePrompt(context),
    parameters: createInvoiceParameters(context),
    actions: {
      invoices: {
        create: true,
      },
    },
  },
  {
    method: 'list_invoices',
    name: 'List Invoices',
    description: listInvoicesPrompt(context),
    parameters: listInvoicesParameters(context),
    actions: {
      invoices: {
        read: true,
      },
    },
  },
  {
    method: 'create_invoice_item',
    name: 'Create Invoice Item',
    description: createInvoiceItemPrompt(context),
    parameters: createInvoiceItemParameters(context),
    actions: {
      invoiceItems: {
        create: true,
      },
    },
  },
  {
    method: 'finalize_invoice',
    name: 'Finalize Invoice',
    description: finalizeInvoicePrompt(context),
    parameters: finalizeInvoiceParameters(context),
    actions: {
      invoices: {
        update: true,
      },
    },
  },
  {
    method: 'retrieve_balance',
    name: 'Retrieve Balance',
    description: retrieveBalancePrompt(context),
    parameters: retrieveBalanceParameters(context),
    actions: {
      balance: {
        read: true,
      },
    },
  },
  {
    method: 'create_refund',
    name: 'Create Refund',
    description: createRefundPrompt(context),
    parameters: createRefundParameters(context),
    actions: {
      refunds: {
        create: true,
      },
    },
  },
  {
    method: 'list_payment_intents',
    name: 'List Payment Intents',
    description: listPaymentIntentsPrompt(context),
    parameters: listPaymentIntentsParameters(context),
    actions: {
      paymentIntents: {
        read: true,
      },
    },
  },
  {
    method: 'list_subscriptions',
    name: 'List Subscriptions',
    description: listSubscriptionsPrompt(context),
    parameters: listSubscriptionsParameters(context),
    actions: {
      subscriptions: {
        read: true,
      },
    },
  },
  {
    method: 'search_documentation',
    name: 'Search Documentation',
    description: searchDocumentationPrompt(context),
    parameters: searchDocumentationParameters(context),
    actions: {
      documentation: {
        read: true,
      },
    },
  },
];

export default tools;

File: python/stripe_agent_toolkit/tools.py
from typing import Dict, List

from .prompts import (
    CREATE_CUSTOMER_PROMPT,
    LIST_CUSTOMERS_PROMPT,
    CREATE_PRODUCT_PROMPT,
    LIST_PRODUCTS_PROMPT,
    CREATE_PRICE_PROMPT,
    LIST_PRICES_PROMPT,
    CREATE_PAYMENT_LINK_PROMPT,
    CREATE_INVOICE_PROMPT,
    CREATE_INVOICE_ITEM_PROMPT,
    FINALIZE_INVOICE_PROMPT,
    RETRIEVE_BALANCE_PROMPT,
    CREATE_REFUND_PROMPT,
    LIST_PAYMENT_INTENTS_PROMPT,
)

from .schema import (
    CreateCustomer,
    ListCustomers,
    CreateProduct,
    ListProducts,
    CreatePrice,
    ListPrices,
    CreatePaymentLink,
    CreateInvoice,
    CreateInvoiceItem,
    FinalizeInvoice,
    RetrieveBalance,
    CreateRefund,
    ListPaymentIntents,
)

tools: List[Dict] = [
    {
        "method": "create_customer",
        "name": "Create Customer",
        "description": CREATE_CUSTOMER_PROMPT,
        "args_schema": CreateCustomer,
        "actions": {
            "customers": {
                "create": True,
            }
        },
    },
    {
        "method": "list_customers",
        "name": "List Customers",
        "description": LIST_CUSTOMERS_PROMPT,
        "args_schema": ListCustomers,
        "actions": {
            "customers": {
                "read": True,
            }
        },
    },
    {
        "method": "create_product",
        "name": "Create Product",
        "description": CREATE_PRODUCT_PROMPT,
        "args_schema": CreateProduct,
        "actions": {
            "products": {
                "create": True,
            }
        },
    },
    {
        "method": "list_products",
        "name": "List Products",
        "description": LIST_PRODUCTS_PROMPT,
        "args_schema": ListProducts,
        "actions": {
            "products": {
                "read": True,
            }
        },
    },
    {
        "method": "create_price",
        "name": "Create Price",
        "description": CREATE_PRICE_PROMPT,
        "args_schema": CreatePrice,
        "actions": {
            "prices": {
                "create": True,
            }
        },
    },
    {
        "method": "list_prices",
        "name": "List Prices",
        "description": LIST_PRICES_PROMPT,
        "args_schema": ListPrices,
        "actions": {
            "prices": {
                "read": True,
            }
        },
    },
    {
        "method": "create_payment_link",
        "name": "Create Payment Link",
        "description": CREATE_PAYMENT_LINK_PROMPT,
        "args_schema": CreatePaymentLink,
        "actions": {
            "payment_links": {
                "create": True,
            }
        },
    },
    {
        "method": "create_invoice",
        "name": "Create Invoice",
        "description": CREATE_INVOICE_PROMPT,
        "args_schema": CreateInvoice,
        "actions": {
            "invoices": {
                "create": True,
            }
        },
    },
    {
        "method": "create_invoice_item",
        "name": "Create Invoice Item",
        "description": CREATE_INVOICE_ITEM_PROMPT,
        "args_schema": CreateInvoiceItem,
        "actions": {
            "invoice_items": {
                "create": True,
            }
        },
    },
    {
        "method": "finalize_invoice",
        "name": "Finalize Invoice",
        "description": FINALIZE_INVOICE_PROMPT,
        "args_schema": FinalizeInvoice,
        "actions": {
            "invoices": {
                "update": True,
            }
        },
    },
    {
        "method": "retrieve_balance",
        "name": "Retrieve Balance",
        "description": RETRIEVE_BALANCE_PROMPT,
        "args_schema": RetrieveBalance,
        "actions": {
            "balance": {
                "read": True,
            }
        },
    },
    {
        "method": "create_refund",
        "name": "Create Refund",
        "description": CREATE_REFUND_PROMPT,
        "args_schema": CreateRefund,
        "actions": {
            "refunds": {
                "create": True,
            }
        },
    },
    {
        "method": "list_payment_intents",
        "name": "List Payment Intents",
        "description": LIST_PAYMENT_INTENTS_PROMPT,
        "args_schema": ListPaymentIntents,
        "actions": {
            "payment_intents": {
                "read": True,
            }
        },
    },
]
